tic_tac_toe/game.py

```diff
@@ -0,0 +1,24 @@
+from tic_tac_toe.board import Board
+from tic_tac_toe.views import CommandLineBoardPresenter
+from tic_tac_toe.user_input import UserInput
+
+
+def run():
+    board = Board()
+    view = CommandLineBoardPresenter()
+    user_selection = UserInput()
+
+    while True:
+        view.display_board(board)
+        try:
+            selection = user_selection.process_input()
+            board.turn(selection)
+        except ValueError:
+            print("Sorry, that input is invalid. Please enter numbers 1 - 9:\n")
+        except IndexError:
+            print("Sorry, that input is out of range. Please enter 1 - 9:\n")
+        except KeyError:
+            print(
+                "Sorry, that position has already been taken, please choose another:\n"
+            )
```

Comment:
There's some nice stuff going on here. I'll let you talk about custom errors with @pedmiston — I think you already recognize that custom error classes would be helpful, and to me this is the biggest example of why: it's a little challenging to understand how these very low-level / primitive errors correspond to the error messages that get printed. Unrelated...

It'd be nice to *wrap* your calls to `print`, for a few reasons:
- all in one place instead of scattered throughout the codebase
- easier to test.
Consider: the `CommandLIneBoardPresenter` doesn't necessarily have to execute the print statement displaying it on the screen... it could just be responsible for formatting the string that gets passed to `print`. That would let you remove the `capfd` calls in that class's tests and simplify it to just comparing two string values.

Speaking of... how will you test this function?

Reply:
@mikeknep @pedmiston Regarding the tests, this seems like a really hard function to test. How do you test a function that doesn't return anything, but just calls other functions and prints stuff?

One thing I'm really struggling with right now is testing any function that takes in user input, prints something, or calls the functions that do. This is the error I've been wrestling with:

`OSError: pytest: reading from stdin while output is captured!  Consider using `-s`.`

I've tried mocking input and disabling capsys, but nothing has worked yet, so I figure the problem must be not understanding how to test this correctly.